File: app/src/main/java/com/pattern/mvvm/view/activity/BaseActivity.java
package com.pattern.mvvm.view.activity;

import android.os.Bundle;
import android.view.WindowManager;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.lifecycle.ViewModel;

public abstract class BaseActivity<DB extends ViewDataBinding, VM extends ViewModel> extends AppCompatActivity {

    protected DB binding;
    protected VM mViewModel;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);

        binding = DataBindingUtil.setContentView(this, getLayoutId());

        /**
         *  Common Place for Activity's View Model Object Creation (Registration)
         * */

        this.mViewModel = mViewModel == null ? getViewModel() : mViewModel;
        binding.setLifecycleOwner(this);

        /**
         *  One Binding Variable for Each XML Layout
         * */

        binding.setVariable(getBindingVariable(), mViewModel);
        binding.executePendingBindings();
    }

    abstract int getLayoutId();

    abstract VM getViewModel();

    abstract int getBindingVariable();

}

File: app/src/main/java/com/pattern/mvvm/view/activity/BaseActivity.kt
package com.pattern.mvvm.view.activity

import android.os.Bundle
import android.view.WindowManager
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.lifecycle.ViewModel

abstract class BaseActivity<DB : ViewDataBinding, VM : ViewModel> : AppCompatActivity() {

    internal lateinit var binding: DB
    internal lateinit var mViewModel: VM

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        window.setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN)

        binding = DataBindingUtil.setContentView(this, layoutId)

        /**
         * Common Place for Activity's View Model Object Creation (Registration)
         */

        this.mViewModel = getViewModel
        binding.lifecycleOwner = this

        /**
         * One Binding Variable for Each XML Layout
         */

        binding.setVariable(bindingVariable, mViewModel)
        binding.executePendingBindings()
    }

    internal abstract val layoutId: Int
    internal abstract val getViewModel: VM
    internal abstract val bindingVariable: Int
}
